### PR TITLE
[base,ottf,test] Refactor ujson ottf console to support spi_device console

### LINKED ISSUE #21726: [manuf,ottf] implement spi_device console over mailbox
Previously, in #17818, a spi_device OTTF console was added to facilitate manufacturing flows that require used of higher speed IO (than is available with UART). When generic mode was removed in #20856, this feature was also removed, since it made use of generic mode.

To facilitate manufacturing flows for Earlgrey PROD, a new spi_device OTTF console should be implemented using the SPI mailbox.

### PR DESCRIPTION
**Pull request by anthonychen1251**
This PR has four commits that:
1. Refactor ujson ottf console and implement `getc` for SPI device.
2. Test the new `getc` implementation.
3. Refactor mem module to support the new SPI device console
4. Add an example test for spi_device ujson ottf console.

This fixes #21726.

### CODE REVIEW

**Comment by anthonychen1251** [comment]
cc @timothytrippel

**Comment by timothytrippel** [comment]
I think you can rebase this on master now since https://github.com/lowRISC/opentitan/pull/24380 has been merged.

File: sw/device/lib/testing/test_framework/ottf_console.c
Comment (timothytrippel) [inline sw/device/lib/testing/test_framework/ottf_console.c:81]: Do you mean:
```suggestion
 * 1. The exact amount of data expected to be sent from the host side must be
```
?

File: sw/device/lib/testing/test_framework/ottf_console.c
Comment (timothytrippel) [inline sw/device/lib/testing/test_framework/ottf_console.c:371]: wouldn't the data lost be `len - written_len`?

File: sw/device/lib/testing/test_framework/ottf_console.c
Comment (timothytrippel) [inline sw/device/lib/testing/test_framework/ottf_console.c:93]: I think instead of `&ottf_console_spi_device` here you want to cast the `io` pointer to the spi_device DIF handle, like is done for the `uart_getc` function.

File: sw/device/tests/BUILD
Comment (timothytrippel) [inline sw/device/tests/BUILD:3279]: lets run this on cw310 too.

File: sw/device/tests/spi_device_ottf_console_test.c
Comment (timothytrippel) [inline sw/device/tests/spi_device_ottf_console_test.c:146]: Add comment describing what this block is testing (since it is similar yet different from the other tests above)? i.e. underlying UJSON OTTF console data pipeline

File: sw/device/tests/spi_device_ujson_ottf_test.c
Comment (timothytrippel) [inline sw/device/tests/spi_device_ujson_ottf_test.c:23]: This test has nothing to do with SW straps correct? If you need some UJSON Structs to send back and for testing, why don't you use this generic `MemRead` test command https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/json/command.h;l=25 , and send MemRead data payloads back and forth for one test: https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/json/mem.h;drc=30d7e787c753caaa03fe68a4a70da1bbcbc1d96f;l=24 (this should test data payloads of greater than 256 bytes).

Then you could also test sending a perso blob ujson struct to and from the host since it is > 4k in size which should excercise multi SPI buffer window transactions.

Also if you could add a block comment to the top describing what you are testing that would be helpful too.

File: sw/device/tests/spi_device_ottf_console_test.c
Comment (anthonychen1251) [reply sw/device/tests/spi_device_ottf_console_test.c:146]: Okay, done.

File: sw/device/lib/testing/test_framework/ottf_console.c
Comment (anthonychen1251) [reply sw/device/lib/testing/test_framework/ottf_console.c]: Ah, you're right. Fixed the typo.

File: sw/device/lib/testing/test_framework/ottf_console.c
Comment (anthonychen1251) [reply sw/device/lib/testing/test_framework/ottf_console.c]: Good point! My original thought was to report the actual data written length along with a data loss error, but it seems to be reasonable to report the length of missing data.

File: sw/device/lib/testing/test_framework/ottf_console.c
Comment (anthonychen1251) [reply sw/device/lib/testing/test_framework/ottf_console.c]: Okay, done.

File: sw/device/tests/BUILD
Comment (anthonychen1251) [reply sw/device/tests/BUILD:3286]: Sure, done.

File: sw/device/tests/spi_device_ujson_ottf_test.c
Comment (anthonychen1251) [reply sw/device/tests/spi_device_ujson_ottf_test.c]: Yes, the test has nothing to do with the functionality of SW straps. I did consider using the generic `Mem*` test commands but decided to use the SW straps command in the end. The tx transaction on SPI device console requires the device to send a sync message to notify the host it's ready for transmission but this limitation doesn't exist on UART console. The `mem` module encapsulates several tx transactions into a single `execute` function. To make the mem module support the workflow SPI device console, the existing `example_mem_ujcmd` test would be affected as well. I've submitted a new commit for refactoring the mem module. Let me know if that works for you.   I also added a test to send a perso blob ujson struct back and forth, thanks for the suggestion!

Comment (anthonychen1251) [reply sw/device/tests/spi_device_ujson_ottf_test.c]: Thanks for the comment. The purpose is to ensure that the rx data has been read  by the host before host starts sending data to the device, so the device won't be blocked by the rx transmission if the read buffer is full and can be ready to receive the tx data. However, after I think about the use case here again, the sync operation in the `execute` seems to be unnecessary if a sync operation has already been performed in advance. Will double confirm on this and see if I can get rid of the `require_sync`

Comment (pamaury) [reply sw/device/tests/spi_device_ujson_ottf_test.c]: Maybe this is a limitation of opentitanlib but I would have thought that the console would have a background thread so it can periodically read the SPI and therefore unblock the device if necessary?

Comment (nbdd0121) [reply sw/device/tests/spi_device_ujson_ottf_test.c]: > The purpose is to ensure that the rx data has been read by the host before host starts sending data to the device, so the device won't be blocked by the rx transmission if the read buffer is full and can be ready to receive the tx data.

Couldn't the host just poll until the rx buffer is empty before sending data out? Would the device be sending data unsolicited to the host?

Comment (anthonychen1251) [reply sw/device/tests/spi_device_ujson_ottf_test.c]: Removed the usage of `require_sync` in the new commit. Host poll until the rx buffer is empty before sending data out should work as well. Ideally, the device should not send unsolicited data to the host. I observed the issue when I tried to log plenty of debug messages on the device side but didn't handle those messages properly on the host side. After checking the behaviors on the existing ottf test code, I think the case I observed is an extreme case that should not occur in the real use scenario. The original purpose is not really necessary now.

**Comment by anthonychen1251** [comment]
Upon pushing new commits, I observed that certain checks, like `CW310 ROM Tests`, were not executed again. Is there a way to rerun all the checks to ensure the latest commits do not break anything?

**Comment by timothytrippel** [comment]
Hmm that is strange. IIRC, closing and re-opening the PR can retrigger CI. Let's try that.

File: sw/device/tests/BUILD
Comment (timothytrippel) [inline sw/device/tests/BUILD:3284]: Can we rename this something like:
```suggestion
    name = "spi_device_ujson_console_test",
    srcs = ["spi_device_ujson_console_test.c"],
```

and likewise the `test_harness` below? 

The current name is confusing since there is also a `spi_device_ottf_console_test` that just tests the SPI device OTTF console, without sending ujson data payloads I believe.

**Review by timothytrippel — APPROVED** [review]

File: sw/device/tests/BUILD
Comment (anthonychen1251) [reply sw/device/tests/BUILD]: Sure, done.

File: sw/device/tests/BUILD
Comment (anthonychen1251) [reply sw/device/tests/BUILD]: Got it. Removed the cw310 part.

**Comment by anthonychen1251** [comment]
The failed test `pattgen_ios_test_fpga_cw310_sival` in the most recent CI pipeline should be unrelated to the commits in the PR.
https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=159102&view=logs&j=5011a995-22d7-5073-e1ce-26fc382a8885&t=875997a6-7e2b-594d-05e6-ce6a085f3383

**Comment by timothytrippel** [comment]
I retriggered CI. Hopefully the transient error should go away. Once it passes, I will merge.

**Pull request merged**
